### PR TITLE
feat: error message on 1M+ amount

### LINKED ISSUE #7485: No alert if payment is over $999.999.99 or under $0.50 -Card input fields are not rendered.
### Description

Stripe (and thus WooPayments) have an upper limit of $999'999.99 per transaction limit.
BUT, there is no error or alert displayed if this is reached within the cart, on checkout, and the payment gateway just becomes inactive. 

### Acceptance criteria

* Show a message where the card details are removed that the payment gateway cannot accept payments of $1'000'000 or more.

### Testing instructions

1. Create a product for $999999.99.
2. Complete checkout. It should succeed.
3. Create product for $1'000'000.
4. Complete checkout.
5. No card details rendered.

### Dev notes

N/A

### Additional context

`Internal: 7135175`

### PR DESCRIPTION
**Pull request by frosso**
Fixes https://github.com/Automattic/woocommerce-payments/issues/7485

#### Changes proposed in this Pull Request

Adds an error notice when the amount exceeds 1M in the settlement currency (EUR, USD, etc.).

#### Testing instructions

- Create a product to the cart to exceed the 1M (e.g.: $1000000 USD or 1000000€)
- Go to the blocks or shortcode checkout
- Select WooPayments as the processor
- The error should appear in place of the payment element component
  - You might notice there is a console error related to Stripe fingerprinting, but I wasn't able to resolve it 😬 
  - I wasn't able to remove the "save payment method" checkbox on blocks checkout

Blocks checkout with SEPA/Card
Screenshot 2024-05-07 at 6 43 06 PM
Screenshot 2024-05-07 at 6 42 56 PM

Shortcode checkout with TT4 theme & SEPA
Screenshot 2024-05-07 at 6 11 29 PM

Shortcode checkout with Storefront theme & SEPA/Card
Screenshot 2024-05-07 at 6 10 58 PM
Screenshot 2024-05-07 at 6 10 52 PM

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8793` or branch name `feat/error-notice-on-1m-amount` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=feat/error-notice-on-1m-amount) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 3e60a91724216b3df5925570e8fa0b1f5bca5548
- Build time: 2024-05-17 08:11:08 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Review by FangedParakeet — COMMENTED** [review]
This is working as expected for me and the changes look fine. 

> I wasn't able to remove the "save payment method" checkbox on blocks checkout

I guess, AFAICT, it appears that this option is set when we register the payment gateway and presumably cannot be updated later. Stripe will only inform us of the error after this fact, so I guess we're stuck here. 

> You might notice there is a console error related to Stripe fingerprinting, but I wasn't able to resolve it 😬

I spent a little time investigating this as well...and was unsurprisingly unable to come up with any more answers than you have. This error does appear to originate from within Stripe's own iframe, so perhaps this is just some error case that they are failing to handle adequately on their end. So it goes.

I did notice this error appearing on the blocks checkout.

paymentRequestPaymentMethod error
_Sad_

I thought this was particularly peculiar, because it seemed to originate from the `paymentRequestPaymentMethod` object, which is part of our payment request button implementation. The error message is unsurprisingly unhelpful, but I noticed that this error still appeared even when I disabled the express checkout methods from my store's payments' settings. 

It looks like [we are registering the payment request button](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/index.js#L156), unless I'm mistaken, in all cases, so perhaps there needs to be additional error handling here--with no fault being given to the changes here nor any onus on resolving this within the current PR. 

I did notice one more thing that did give me a little further pause for thought. In both our blocks and shortcode checkouts, when the payment error returns an error and is unusable, the "Place order" button is still selectable.

On the blocks checkout, if we select the "Place order" button, we are returned this error message, which is somewhat meaningful and prevents the checkout from continuing.

Blocks checkout error message
_Blocks checkout place order error message_

However, on the shortcode, we are presented with this lovely "r is undefined" error message instead.

Shortcode checkout error message
_Wut_

I was wondering whether you had encountered this rather unattractive message or if you had any thoughts about how we might deal with it? Also, did you consider disabling the "Place order" button at all, in case there are no other payment gateways available. I'm not sure if there is a way for a payment gateway to inform WC that the gateway is inoperable, but I'd presume there isn't one. 

Interested to hear your thoughts on the above before proceeding. 🙏

**Comment by frosso** [comment]
> However, on the shortcode, we are presented with this lovely "r is undefined" error message instead.

Ah, good find! While debugging this, I noticed that the exception is being thrown [around here](https://github.com/Automattic/woocommerce-payments/blob/f7bbd327be1b48bbca5e1c82f996e9ad3870dcd6/client/checkout/classic/payment-processing.js#L519) on shortcode checkout.

But it's weird that blocks checkout [uses the same function](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/payment-processor.js#L172), but a different error is thrown 🙄 

I'll continue looking into this, thanks for pointing it out! 👍

**Comment by frosso** [comment]
> I'll continue looking into this, thanks for pointing it out! 👍

@FangedParakeet in 5f15aa63ef6c0b554badd28a1bbedcf32eade287 I made a change to align the implementation with the blocks checkout a bit more - on the blocks checkout [we rely on the `event.complete` attribute](https://github.com/Automattic/woocommerce-payments/blob/5f15aa63ef6c0b554badd28a1bbedcf32eade287/client/checkout/blocks/payment-processor.js#L241-L241) [on the `onChange` handler](https://github.com/Automattic/woocommerce-payments/blob/5f15aa63ef6c0b554badd28a1bbedcf32eade287/client/checkout/blocks/payment-processor.js#L260-L260). [So that's why the error was different](https://github.com/Automattic/woocommerce-payments/blob/5f15aa63ef6c0b554badd28a1bbedcf32eade287/client/checkout/blocks/payment-processor.js#L143-L143).

Now the two implementations should be a little bit more aligned.

File: client/checkout/classic/payment-processing.js [hunk 1]
```diff
@@ -400,6 +401,27 @@ export async function mountStripePaymentElement( api, domElement ) {
 		gatewayUPEComponents[ paymentMethodType ].upeElement ||
 		( await createStripePaymentElement( api, paymentMethodType ) );
 	upeElement.mount( domElement );
+	upeElement.on( 'change', ( e ) => {
+		gatewayUPEComponents[ paymentMethodType ].isPaymentInformationComplete =
+			e.complete;
+	} );
+	upeElement.on( 'loaderror', ( e ) => {
```
Comment (FangedParakeet) [inline client/checkout/classic/payment-processing.js:408]:
```suggestion
	upeElement.on( 'change', ( event ) => {
		gatewayUPEComponents[ paymentMethodType ].isPaymentInformationComplete =
			event.complete;
	} );
	upeElement.on( 'loaderror', ( event ) => {
```
[Just sayin'.](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#naming-conventions) 🤷

**Review by FangedParakeet — APPROVED** [review]
Nice one! Tested on blocks and shortcode, plus a few checkouts with new and saved information just to be safe, and this is working mostly as expected for me. I just want to raise two small peculiarities that I noticed along the way.

Firstly, on my hideous dark-mode test-site, I noticed that the warning was sadly illegible. This isn't necessarily the fault of this PR and I'm not sure if there's much in our power to improve this, so don't think it's a big deal. 

Illegible warning
_I can't see_

Also, I noticed on the shortcode checkout, if I tried to place an order with the payment element prevented from working, the error at the top of the page appeared as expected, but the warning on the payment element disappeared, leaving the gateway empty. On the block checkout, both warnings persisted.

Empty gateway
_Sad, lonely gateway_

This also isn't really a big deal, because the goal is to prevent the checkout and to show a warning--and one warning is better than none. 

Anyways, I don't think either of these problems are showstoppers, so let's keep the train rolling. :shipit:

File: client/checkout/classic/payment-processing.js (hunk 1 above)
Comment (frosso) [reply client/checkout/classic/payment-processing.js:408]: Unfortunately, `event` is already defined a few lines above, so I can't name this the same :( 

https://github.com/Automattic/woocommerce-payments/blob/feat/error-notice-on-1m-amount/client/checkout/classic/payment-processing.js#L392

**Comment by frosso** [comment]
> Firstly, on my hideous dark-mode test-site, I noticed that the warning was sadly illegible. This isn't necessarily the fault of this PR and I'm not sure if there's much in our power to improve this, so don't think it's a big deal.

Do you happen to know if the other WC frontend notices that appear on your site also have insufficient color contrast?

**Comment by FangedParakeet** [comment]
> Do you happen to know if the other WC frontend notices that appear on your site also have insufficient color contrast?

Good point! This main notice also requires some squinting to read too, so likely this is just a result of me making my site as hideous as humanly possible. 😂 

Illegible notice
👍

**Comment by csmcneill** [comment]
I tested this a bit today and have these observations:

1. The error also shows up on the My account > Payment methods page.

CleanShot 2024-05-20 at 13 33 28@2x

2. The error notice does not seem to account for some currency-specific formatting:

- For some currencies, I'd expect for the value in the error to be non-decimal and use the same separators, abbreviations, decimal places, and currency symbols

CleanShot 2024-05-20 at 13 34 37@2x

CleanShot 2024-05-20 at 13 58 42@2x

- Affected currencies included but are not limited to: ₡ CRC, $ CLP, Fr DJF, Fr GNF, Fr KMF, ₲ PYG, Ush UGX, ₫ VND, CFA XOF, ₪ ILS, and Fr XPF
- I thought that this affected all zero-decimal currencies, but some of them work just fine because of their standard format: ¥ JPY, ₩ KRW, Ar MGA, Fr RWF, and Vt VUV

CleanShot 2024-05-20 at 13 35 59@2x

**Comment by csmcneill** [comment]
I did some more testing and found another quirk: saved cards can still be used for orders exceeding the maximum amount, but errors aren't always shown.

On my self-hosted site, using a saved payment via the Checkout Block produces a `400` response. This generates an error on the front-end:

CleanShot 2024-05-20 at 19 23 05@2x

However, on my Atomic site, using a saved payment via the Checkout Block produces a `402` response. This generates no error on the front-end _at all_ because 402 errors are reserved for WP Cloud on Atomic p9F6qB-cgE-p2 🫠

**Comment by frosso** [comment]
Hey @csmcneill , thanks for testing! 👋 
What version of WC were you using for your tests? [This PR](https://github.com/woocommerce/woocommerce/pull/45979)
 (to ensure a `400` is returned instead of `402`) should be in one of the latest betas, and it should appear in one of the next releases.

**Comment by csmcneill** [comment]
@frosso That's it! I was testing the latest Nightly build on my self-hosted site. I forgot to revert back to WooCommerce 8.9.x before testing this :)
